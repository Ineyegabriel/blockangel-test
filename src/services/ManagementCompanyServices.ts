import { CompanyData } from "../init/CompanyData";
import {
  CompanyPayments,
  ContactPayment,
  PaymentStatus,
} from "../model/ContactPayment";
import { ManagementCompany } from "../model/ManagementCompany";

export class ManagementCompanyServices {
  private companyData: CompanyData;

  constructor() {
    this.companyData = new CompanyData();
  }

  // Test:
  // Implement logic outlined in the 3 functions below. Feel free to add any comments to support your
  // implementation. Don't spend any more than 90 mins on this
  //
  // Note: Don't change the function signature (including async)
  //

  // 1. Create a Map:
  //      From: this.companyData.getCompaniesData()
  //      Key: company.sequence
  //      Value: company
  //
  //  Return the Map
  public async getCompanySequenceMap() {
    //TODO - Implement
    let managementCompanies: ManagementCompany[] = [];
    const companySequenceMap: Map<number, ManagementCompany[]> = new Map();

    await this.companyData.getCompaniesData().then((data) => {
      managementCompanies = data;
      // Map companies to sequence map
      managementCompanies.forEach((company) => {
        if (company.sequence !== undefined) {
          // Add to map
          if (companySequenceMap.has(company.sequence)) {
            companySequenceMap.get(company.sequence)?.push(company);
          } else {
            // Create new array
            companySequenceMap.set(company.sequence, [company]);
          }
        }
      });
    });
    return companySequenceMap;
  }

  // 2. This function extracts the company related to a given contact Id
  //      Input: contactId, in the format of ss-xxxx where ss represents the company sequence (see ManagementCompany)
  //      Output: company identified by ss
  //
  // Hint: Use 1. getCompanySequenceMap()
  public async getCompanyRelatedToContactId(contactId: String) {
    //TODO - Implement
    const companySequence = Number(contactId.split("-")[0]);
    const companyMap = await this.getCompanySequenceMap();
    return companyMap.get(companySequence);
  }

  // 2.This function processes contact payments to create equivalent company payments that are filtered and sorted.
  //
  //  A. Create a new Model object: CompanyPayment, with same fields as ContactPayment with addition of companyId and
  //     company Name. Process contact payments (this.companyData.getContactPaymentsData()) to create CompanyPayments array
  //     enriched with companyId and company name.
  //
  //     Note: use map to process contact payments, contactPayments.map(....
  //
  //  B. Filter the array of CompanyPayments by PaymentStatus.PAID
  //
  //  C. Sort the CompanyPayments by date
  //
  //  Return CompanyPayments array
  //
  // Hint:
  //      - in mapping you will need to use this.getCompanyRelatedToContactId
  //      - in map logic you might run into promise issues, Promise.all!!
  //
  public async processContactPayments() {
    //TODO - Implement
    const contactPaymentsData = await this.companyData.getContactPaymentsData();

    // A. Create CompanyPayments array enriched with companyId and companyName
    const companyPayments: Array<CompanyPayments> = await Promise.all(
      contactPaymentsData.map(async (contactPayment) => {
        const company: ManagementCompany | undefined = (
          await this.getCompanyRelatedToContactId(contactPayment.contactId)
        )?.[0];
        return {
          ...contactPayment,
          companyId: company?.id || "",
          companyName: company?.companyName || "",
        };
      })
    );

    // B. Filter by PaymentStatus.PAID
    const paidCompanyPayments = companyPayments.filter((cp) => {
      if (cp.status !== PaymentStatus.PAID) {
        return false;
      }

      return true;
    });

    // C. Sort by date
    paidCompanyPayments.sort((a, b) => a.date - b.date);

    return paidCompanyPayments;
  }
}
