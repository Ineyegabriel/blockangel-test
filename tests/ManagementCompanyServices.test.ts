import { ManagementCompanyServices } from "../src/services/ManagementCompanyServices";

describe("Management Company Services", () => {
  beforeEach(() => {
    const mockedSystemTime = "2023-08-10T01:36:37.675Z";
    jest.useFakeTimers().setSystemTime(new Date(mockedSystemTime));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  test("getCompanySequenceMap returns a map of company sequence and company details", async () => {
    let mcServices = new ManagementCompanyServices();
    const action = await mcServices.getCompanySequenceMap();
    expect(action).toMatchSnapshot();
  });

  test("processContactPayments returns a map of CompanyPayments sorted by payment status", async () => {
    let mcServices = new ManagementCompanyServices();
    const action = await mcServices.processContactPayments();

    expect(action).toMatchSnapshot();
  });
});
