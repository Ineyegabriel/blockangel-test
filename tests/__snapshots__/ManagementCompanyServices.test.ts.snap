// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Management Company Services getCompanySequenceMap returns a map of company sequence and company details 1`] = `
Map {
  1 => [
    {
      "companyName": "aCompany",
      "id": "1",
      "sequence": 1,
    },
  ],
  3 => [
    {
      "companyName": "zCompany",
      "id": "3",
      "sequence": 3,
    },
    {
      "companyName": "xCompany",
      "id": "3",
      "sequence": 3,
    },
  ],
  2 => [
    {
      "companyName": "bCompany",
      "id": "2",
      "sequence": 2,
    },
  ],
}
`;

exports[`Management Company Services processContactPayments returns a map of CompanyPayments sorted by payment status 1`] = `
[
  {
    "amount": 10,
    "companyId": "2",
    "companyName": "bCompany",
    "contactId": "02-0001",
    "date": 1691631397672,
    "status": "PAID",
  },
  {
    "amount": 10,
    "companyId": "2",
    "companyName": "bCompany",
    "contactId": "02-0001",
    "date": 1691631397672,
    "status": "PAID",
  },
  {
    "amount": 20,
    "companyId": "2",
    "companyName": "bCompany",
    "contactId": "02-0003",
    "date": 1691631397674,
    "status": "PAID",
  },
]
`;
